### ISSUE TITLE
"errorMessage": "Unable to import module 'otel_wrapper': No module named 'utils'"

### ISSUE THREAD

**Issue by itaybaror** [open-telemetry/opentelemetry-lambda-extension#254]
### Discussed in https://github.com/open-telemetry/opentelemetry-lambda/discussions/253

<div type='discussions-op-text'>

<sup>Originally posted by **itaybaror** June 22, 2022</sup>
I have been trying to auto instrument my python lambda function using the following link.

https://aws-otel.github.io/docs/getting-started/lambda/lambda-python

However i get the following error:

![Screen Shot 2022-06-22 at 3 17 56 PM](https://user-images.githubusercontent.com/107562167/175118944-36d61ab8-1ac2-4ec7-94f2-d2911553462b.png)

After a lot of playing around, this only happens when i have the AWS_LAMBDA_EXEC_WRAPPER env variable set. Without it, the lambnda function tests with no problems.

I ran a smaller sample project i found in a blog article and everything worked. So, what in my project (which has another layer), could be causing this error?

Please help i have been stuck on this for a while.

</div>

**Comment by UniqueScreenName** [comment]
I am seeing the same problem:
 {
"errorMessage": "Unable to import module 'otel_wrapper': No module named 'fss_request_xml_mapper'", 
"errorType": "Runtime.ImportModuleError", 
"stackTrace  ": []
}

In this case, the "fss_request_xml_mapper" module is found in another layer used by the function.  I  have followed the steps outlined in https://aws-otel.github.io/docs/getting-started/lambda/lambda-python as well.  The WARNING seen in the reporters log is also in mine as well.

**Comment by bryan-aguilar** [comment]
Can you please try to use the latest python arns to see if the error persists?

**Comment by UniqueScreenName** [comment]
No, that didn't seem to make any difference using the latest arn from 
https://aws-otel.github.io/docs/getting-started/lambda/lambda-python

Same error.

**Comment by TheNomet** [comment]
Hi @UniqueScreenName, I think your problem is to be found in how you package your other layer.
I would suggest
1. use `sam cli` for deploying the entire infrastructure, and let sam package the layer for you
2. if you want to create the layer yourself, install your `utils` library in `my_library/python` and then `cd my_library && zip -r ../utils.zip *. You can now try to upload your layer and see what happens.

**Comment by myles2007** [comment]
I recently experienced the same error message while converting an existing project to use OpenTelemtry. 

After introducing the layer, I started receiving an error message like:

```
[ERROR] Runtime.ImportModuleError: Unable to import module 'otel_wrapper': No module named 'package1/package2/module.hander'
```

My issue turned out to be a difference in how the standard Lambda runtime and the `otel_wrapper.py` script—specifically the Lambda auto instrumentation—import handlers. In short, slash-based paths result in an error above during instrumentation. The wrapper itself and the standard runtime handle this format (`package1/package2/module.handler`), but the Lambda instrumentation does not.

This specific issue can be easily worked around by converting the handler target to use only dots:

 - `package1.package2.module.handler`

I created https://github.com/myles2007/OpenTelemIssue-PythonLambdaLayer-v1.14.0 to further explain and demonstrate the problem. I intend to create a specific Issue tomorrow (and hopefully a fix soon after).

**Comment by myles2007** [comment]
For anyone who may come across this issue and find themselves in the scenario I outlined above, the issue I created is [#1465](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1465).

**Comment by jkclark** [comment]
> My issue turned out to be a difference in how the standard Lambda runtime and the `otel_wrapper.py` script—specifically the Lambda auto instrumentation—import handlers. In short, slash-based paths result in an error above during instrumentation. The wrapper itself and the standard runtime handle this format (`package1/package2/module.handler`), but the Lambda instrumentation does not.
> 
> This specific issue can be easily worked around by converting the handler target to use only dots:
> 
> * `package1.package2.module.handler`


I would never ever have solved this problem without your comment, which I'm pretty sure is the only such comment in existence. Thank you so much @myles2007

**Comment by myles2007** [comment]
@jkclark
You're welcome! I'm glad to hear it helped you out. :)

For a little added context in case it helps you or anyone else out, I encountered this issue while using [SST](https://sst.dev/). Previously this detail was just happenstance—it was quite possible to fix by using the dot-based notation I previously described—so I omitted it.

Unfortunately, as of SST 2.x, the OpenTelemetry layer auto instrumentation and SST are now at odds with each other. SST requires slash-based path notation and the OpenTelemetry layer still requires the dot-based notation. Short of modifying SST or the the layer, I was unable to resolve this issue.

Long story short, the root of this issue is that SST now infers your source path (i.e., the directory to package up):

<img width="1001" alt="image" src="https://user-images.githubusercontent.com/235987/227264958-7c483884-98bc-4962-b734-c662929b4805.png">

([source](https://docs.sst.dev/upgrade-guide#:~:text=We%27ve%20removed%20the%20need%20for%20srcPath%20in%20function%20definitions%20but%20all%20your%20handler%20paths%20need%20to%20be%20specified%20relative%20to%20the%20root%20of%20the%20project.))

For Python, it does this based on a search up your path to the nearest requirements file ([one of a few types](https://github.com/serverless-stack/sst/blob/032bfb7e5a10937523e63de03b1cddae7027e647/packages/sst/src/runtime/handlers/python.ts#L31-L37)). An assumption is made that the path looks like a file system path ([source](https://github.com/serverless-stack/sst/blob/032bfb7e5a10937523e63de03b1cddae7027e647/packages/sst/src/util/fs.ts#L5-L12)) for the purposes of traversal.

While this _could_ be resolved in SST, a framework's choice of how the handler path is specified is up to that framework (in my opinion). Ultimately, I still believe this is something that should be addressed in the OpenTelemetry layer as it is intended to be something you can introduce seamlessly into your Lambda, but it's not compatible with the options Lambda provides for handler specification in a Python runtime.

**Comment by sebas1017** [comment]
Thank you very much, I had the same error and your solution was the only salvation.  @myles2007  
To solve it, reducing side effects, modify the otel_wrapper.py file of the otel-lambda layer, and... as you see on line 480, modify self.wrapped_module_name and apply a replace to eliminate / and leave it alone., also add a try except to this instrument function since at least it could have a trace of errors in cloudwatch :) and it worked perfectly

![image](https://github.com/open-telemetry/opentelemetry-lambda/assets/54646213/2ccbb354-3790-4ce2-af5f-73a0a2772ce7)

**Comment by myles2007** [comment]
You're welcome, @sebas1017! I'm happy to hear my comments were helpful to you. :)

Hopefully https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1465 can get some traction one of these days. I suppose I should consider just submitting a fix myself, but I've been a little hesitant of spending the time given the lack of even a reply on my issue. 😞

**Comment by brett-bim** [comment]
Has there ever been any movement on this?  Currently getting this while trying to integrate the otel layer into an existing lambda with a custom built layer.  We do not use / based function handlers so i don't think it's the same problem as @myles2007.  

`Runtime.ImportModuleError: Unable to import module 'otel_wrapper': No module named 'paramiko'`

We have a layer called `paramiko` that this lambda uses. I sense that the `otel_wrapper` is trying to interact with it in some way but not sure how.

**Comment by tylerbenson** [comment]
@myles2007 regarding your comment:
>  I suppose I should consider just submitting a fix myself, but I've been a little hesitant of spending the time given the lack of even a reply on my issue. 😞

I apologize for the lack of response in this repo.  The FAAS SIG has had a lot of staffing changes over the past year.  We aren't really staffed for a lot of troubleshooting right now. (If folks want to be more involved, we'd love to have you.)

If you would like to submit a PR to fix this, I'd be happy to get folks to review. (If in the contrib repo, I can reach out to the maintainer to encourage reviews.)

**Comment by forbushbl** [comment]
It appears that there are two issues that result in a similar error message:

### Issue 1: Lambda dependencies in the `python/lib/python3.x/site-packages` directory
AWS Lambda supports attaching a layer containing code dependencies to your Lambda function. According to the [documentation](https://docs.aws.amazon.com/lambda/latest/dg/python-layers.html#python-layer-paths) you can place dependencies in either the `python` or the `python/lib/python3.x/site-packages` directory, however opentelemetry-python causes your function to break if your dependencies are in `python/lib/python3.x/site-packages`. 

**Easy Workaround:** Add the PYTHONPATH environment variable to your Lambda function with the value `/opt/python/lib/python3.x/site-packages`. Replace `python3.x` with the runtime you're using (Example: `/opt/python/lib/python3.13/site-packages`)

**Harder Workaround:** Repackage your Lambda's dependencies and put them in the `python` directory. This was undesirable for me because of the sheer number of Lambda's I'm dealing with. Also, I suspect this was more of an oversight by the opentelemetry-lambda project and I'm hoping it gets fixed at some point. 🤞

### Issue 2: Using `/` in the handler 
AWS Lambda supports using `/` or `.` when you specify the path to the Python function within your code that the Lambda service should invoke. However, opentelemetry-lambda breaks when you use `/`.

Example:
- `util/math/lambda_function.lambda_handler` (broken ❌)
- `util.math.lambda_function.lambda_handler` (works! ✅)

**Workaround:** Update your Lambda to use `.` in the handler field.
